File: Zaucy/Zaucy/Clientzaucy/zaucy/lib/Entities/Coupon.dart
import 'dart:convert';

class Coupon {
  final int id;
  final String code;
  final DiscountType discountType;
  final double discountValue;
  final Category category;
  final DateTime expiryDate;
  final int usageCount;
  final CouponStatus status;
  final DateTime createdAt;
  final double? minimumOrderValue;
  final bool? isFirstTimeUserOnly;

  Coupon({
    required this.id,
    required this.code,
    required this.discountType,
    required this.discountValue,
    required this.category,
    required this.expiryDate,
    required this.usageCount,
    required this.status,
    required this.createdAt,
    this.minimumOrderValue,
    this.isFirstTimeUserOnly,
  });

  static Future<Coupon> fromJson(Map<String, dynamic> json) async {
    return Coupon(
      id: json['id'],
      code: json['code'],
      discountType: DiscountType.values.byName(json['discountType']),
      discountValue: (json['discountValue'] as num).toDouble(),
      category: Category.values.byName(json['category']),
      expiryDate: DateTime.parse(json['expiryDate']),
      usageCount: json['usageCount'],
      status: CouponStatus.values.byName(json['status']),
      createdAt: DateTime.parse(json['createdAt']),
      minimumOrderValue: json['minimumOrderValue'] != null
          ? (json['minimumOrderValue'] as num).toDouble()
          : null,
      isFirstTimeUserOnly: json['isFirstTimeUserOnly'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'code': code,
      'discountType': discountType.name,
      'discountValue': discountValue,
      'category': category.name,
      'expiryDate': expiryDate.toIso8601String(),
      'usageCount': usageCount,
      'status': status.name,
      'createdAt': createdAt.toIso8601String(),
      'minimumOrderValue': minimumOrderValue,
      'isFirstTimeUserOnly': isFirstTimeUserOnly,
    };
  }
}

enum DiscountType { PERCENTAGE, FLAT }

enum Category { VEG, NON_VEG, ALL }

enum CouponStatus { ACTIVE, INACTIVE, DISCARDED }

File: Zaucy/Zaucy/Zaucy/src/main/java/com/Pizza/Zaucy/Beans/Coupon.java
package com.Pizza.Zaucy.Beans;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "coupons")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Coupon {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true)
    private String code; // Unique coupon code

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private DiscountType discountType; // PERCENTAGE or FLAT

    @Column(nullable = false)
    private BigDecimal discountValue; // Discount amount or percentage

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Category category; // VEG, NON_VEG, ALL

    @Column(nullable = false)
    private LocalDateTime expiryDate; // Expiry date

    @Column(nullable = false)
    private Integer usageCount; // How many times it can be used

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private CouponStatus status; // ACTIVE, INACTIVE, DISCARDED

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt = LocalDateTime.now();
    @Column(nullable = true)
    private BigDecimal minimumOrderValue; // Minimum order value to use the coupon (nullable)

    @Column(nullable = true)
    private Boolean isFirstTimeUserOnly;

    public enum DiscountType {
        PERCENTAGE, FLAT
    }

    public enum Category {
        VEG, NON_VEG, ALL
    }

    public enum CouponStatus {
        ACTIVE, INACTIVE, DISCARDED
    }
}
